File: examples/sdjas.py
#
# $Id$
#
# adapted from the Symbolicdata and Sage code
#
# communicate with the SPARQL endpoint (http request)
#import requests
import httplib, urllib
# easily handle URLs
from os.path import split as pathsplit
from urlparse import urlsplit
# for nice output of dictionaries: json.dumps(dict, indent = 4)
# mostly for debugging reasons (will be removed later)
# for old jython versions it must be loaded from an 
# external library jyson-1.0.2.jar
#import json 
from com.xhaus.jyson import JysonCodec as json
# parse the sd.ini file
from ConfigParser import SafeConfigParser
# parse the xml of the resource files
from xml.dom.minidom import parse, parseString
# output lists nicely (there might be a better way)
from textwrap import wrap as textwrap
# not needed
#from jas import *

# Some internal helper functions that are not meant to be
# called by the user
def _uri_to_name(uri):
    """
    Converts a uri to a name or key by only taking everything
    after the last / or (if present) #.

    Examples:
        http://example.com/test             ->   test
        http://example.com/model#testedBy   ->   testedBy
    """
    usplit = urlsplit(uri)
    if usplit.fragment != '':
        return usplit.fragment
    else:
        return pathsplit(usplit.path)[-1]

def _pprint(l):
    """
    Formats a list l to be displayed in a tabular layout. It is
    possible to pass an integer width to the textwrap function.
    The width of the terminal window could be obtained via the
    Python console module. However, since it is not included
    in Jas, we decided not to use it. The default width that
    textwrap uses is set to 70. There might be a better way to
    do this.
    """
    col = max([len(x) for x in l]) + 3
    padded = ''.join([x.ljust(col) for x in l])
    print '\n'.join(textwrap(padded))

def get_value_for_URI(sd, URI, predicate):
    """
    A quick convienience function to retrieve a single value
    of a given triple (object, predicate, ...)

    The parameter sd is a SymbolicData object that contains
    information about the SPARQL endpoint.
    """
    result = None
    query = '''SELECT * WHERE { <%s> <%s> ?x }''' % (URI, predicate)
    try:
        result = SPARQL(sd, query).json['results']['bindings'][0]['x']['value']
    except:
        pass
    return result



# Class definitions start here
class SymbolicData:
    """
    Access to the database of ideals as provided by the
    SymbolicData Project (http://symbolicdata.org).
    """

    def __init__(self, sparql = 'symbolicdata.org'):
        """
        The constructor parses the sd.ini file and sets up some variables.
        An optional parameter can be passed to select the SPARQL endpoint
        that should be used. The keywords for different SPARQL endpoints are
        defined in the sd.ini file.

        The default SPARQL endpoint is the one from symbolicdata.org
        """
        self._sparql = sparql
        self._ideals = None
        self._parser = SafeConfigParser()
        self._parser.read(['sd.ini','examples/sd.ini'])
        self.sd = self._parser.get('symbolicdata', 'sd')
        try:
            self.url = self._parser.get('sparql', self._sparql)
        except:
            raise ValueError("The SPARQL endpoint referenced by '%s' was not found in the sd.ini file." % self._sparql)
        self.sdhost = self._parser.get('DEFAULT', 'sdhost')
        self.sqpath = self._parser.get('sparql', 'path')
        #print "SymbolicData() initialized"
        #print "url    = " + str(self.url)
        #print "sdhost = " + str(self.sdhost)

    def get_ideals(self, force_reload = False):
        """
        Returns a Python list of ideals.
        """
        if self._ideals == None or force_reload == True:
            self.list_ideals(False, force_reload)
        return self._ideals

    def list_ideals(self, output = True, force_reload = False):
        """
        Lists all the available ideals.
        """
        if self._ideals == None or force_reload == True:
            r = SPARQL(self, self._parser.get('queries', 'list_ideals'))
            self._ideals = [_uri_to_name(x['ideal']['value']) for
                      x in r.json['results']['bindings']]
        if output:
            _pprint(self._ideals)

    def get_ideal(self, uri):
        """
        Returns an ideal as a Jas object that is ready to be used by
        Jas.
        """
        return SD_Ideal(self, uri).get_ideal();

    def get_sd_ideal(self, uri):
        """
        Returns an internal object that represents the SymbolicData
        database object. (See below for details)
        """
        return SD_Ideal(self, uri)



class SPARQL:
    """
    This is a 'wrapper' class for SPARQL queries. A class might be
    a slight overkill. It was made with the idea, that one can store
    the query and the result together, to re-evaluate both without
    having to access the server. However, in the end this feature
    was not really needed.
    """
    def __init__(self, sd, query, output = 'json'):
        self._sd = sd;
        self._query = query;
        self._data = {
            'output' : output,
            'query' : query
        }
        #self.response = requests.get(self._sd.url, params = self._data)
        #print "url = " + str(self._sd.url)
        conn = httplib.HTTPConnection(self._sd.url)
        #print "conn = " + str(conn)

        #print "query = " + str(query)
        _path = self._sd.sqpath + "?" + urllib.urlencode(self._data)
        #print "path = " + str(_path)
        conn.request("GET", _path );
        self.response = conn.getresponse();
        if self.response.status != 200:
           print self.response.status, self.response.reason, "\n"
           raise IOError, "HTTP GET %s not successful" % _path

        self.head = self.response.msg
        #print "head = " + str(self.head)
        self.text = self.response.read()
        #print "body = " + str(self.text)
        self.json = json.loads(self.text)
        #print "json = " + str(self.json)
        conn.close()


class SD_Ideal:
    """
    This class represents a SymbolicData database object. The
    constructor takes a complete URI or a name SUBJ (the latter of which
    will be prefixed with the 'ideal' value from the sd.ini)

    Any triple of the form (SUBJ, PRED, OBJ) will yield a field PRED*
    for the SD_Ideal object with the value OBJ, where PRED* is the
    ending piece of PRED URI as defined by the function _uri_to_name()

    A SPARQL endpoint is needed. As a future improvement, it could be
    nice to directly parse an RDF in a convienient serialization.
    """

    def __init__(self, sd, name):
        """
        sd is a SymbolicData object, the name can be a complete URI or shortened
        name as defined by _uri_to_name(). The latter will be prefixed with the
        'ideal' value from the sd.ini. Namespaces like "sd:Wu-90" are not
        (yet) supported.

        Appart from retrieving the information from the SPARQL endpoint, the
        resource data (XML files) is needed as well. While the SPARQL endpoint
        can be substituted by another SPARQL endpoint, the links to the resource
        files are 'hard-coded' into the RDF data. The possibility to use a
        (possibly 'hand-filled') cache will be included in the next update.
        """
        self._sd = sd
        # quick test, if the given name already is an uri
        if name[:7] == 'http://':
            self.uri = name
        else:
            self.uri = "%s%s" % (self._sd._parser.get("symbolicdata", "ideal"), name)

        self.hasXMLResource = False
        self.hasLengthsList = ''
        self.hasDegreeList = ''
        self.hasParameters = ''

        # we set up the query to get all predicate values
        # of the URI/polynomial system/ideal
        query = '''
            PREFIX sd: <%s>
            SELECT ?p ?o WHERE {
                <%s> ?p ?o
            }''' % (self._sd.sd, self.uri)
        self._request = SPARQL(self._sd, query)

        if len(self._request.json['results']['bindings']) == 0:
            raise ValueError("No data found for <%s>.\nMaybe the name was misspelled or the SPARQL endpoint is unavailable." % self.uri)

        # append the keys to the self.__dict__.
        for t in self._request.json['results']['bindings']:
            uri = t['p']['value']
            obj = t['o']['value']
            self.__dict__[_uri_to_name(uri)] = obj

        # Next we need a resource file with the actual expressions that are
        # used to generate the ideal.
        #
        # There are four cases that need to be dealt with
        #     (1) the ideal is constructed direclty
        #         from an IntPS with related XML resource
        #     (2) the ideal is a flat variant of another
        #         ideal
        #     (3) the ideal is obtained by homogenizing
        #         another ideal
        #     (4) the ideal is obtained by parameterizing another
        #         ideal
        # Please note: While it might seem that only one of (2) and (4)
        # should be included, both are needed to map the actual history
        # of how these ideals were obtained.

        # case 1
        if 'relatedPolynomialSystem' in self.__dict__.keys():
            self.__addXMLResource(get_value_for_URI(self._sd, self.relatedPolynomialSystem, self._sd.sd+'relatedXMLResource'))
            self.hasXMLResource = True
            #print "relatedPolynomialSystem " + str(name)

        # case 2
        if 'flatten' in self.__dict__.keys():
            parent_name = self.flatten
            parent = SD_Ideal(self._sd, parent_name)
            self.variablesCSV = self.hasVariables
            self.variables = map(lambda x: str(x).strip(), self.variablesCSV.rsplit(","))
            self.basis = parent.basis
            #print "flatten " + str(parent_name) + ", name = " + str(name)

        # case 3
        if 'homogenize' in self.__dict__.keys():
            parent_name = self.homogenize
            if 'homogenizedWith' in self.__dict__.keys():
                hv = self.homogenizedWith
                parent = SD_Ideal(self._sd, parent_name)
                self.variablesCSV = parent.variablesCSV + "," + hv
                self.variables = parent.variables
                self.variables.append(hv)
                self.basis = parent.jas_homogenize(hv)
            #print "homogenize " + str(parent_name) + ", name = " + str(name)

        # case 4
        if 'parameterize' in self.__dict__.keys():
            parent_name = self.parameterize
            parent = SD_Ideal(self._sd, parent_name)
            self.variablesCSV = self.hasVariables
            self.variables = map(lambda x: str(x).strip(), self.variablesCSV.rsplit(","))
            self.basis = parent.basis
            #print "parameterize " + str(parent_name) + ", name = " + str(name)

        # now we got the variables, the parameters and
        # the strings/expressions for the polynomials
        self.__constructJasObject()

    def get_ideal(self):
        """
        Return the ideal as a Jas objects.
        """
        #return ideal(self.sageBasis)
        #print "jasRing = " + str(self.jasRing)
        return self.jasRing.ideal(list=self.jasBasis)

    def __addXMLResource(self, link):
        #xml = requests.get(link).text
        #print "link = " + str(link)
        #url = link[0:23]
        path = link[23:] # hack for lost domain
        #print "url = " + str(url)
        #url = self._sd.url[:-5]
        url = self._sd.sdhost
        #print "url = " + str(url)
        conn = httplib.HTTPConnection(url)
        #print "conn = " + str(conn)
        #print "path = " + str(path)
        conn.request("GET", path );
        xml = conn.getresponse().read();
        print _uri_to_name(link) + " = " + str(xml)

        xmlTree = parseString(xml)

        # Code snipped borrowed from Albert Heinle
        if (xmlTree.getElementsByTagName("vars") == []): # Check, if vars are there
            raise IOERROR("The given XMLString does not contain variables for the IntPS System")
        if (xmlTree.getElementsByTagName("basis") == []): # Check, if we have a basis
            raise IOERROR("The given XMLString does not contain a basis for the IntPS System")
        # -------------------- Input Check finished --------------------
        # From here, we can assume that the input is given correct
        self.variablesCSV = (xmlTree.getElementsByTagName("vars")[0]).firstChild.data
        self.variables = map(lambda x: str(x).strip(), self.variablesCSV.rsplit(","))
        polynomials = xmlTree.getElementsByTagName("basis")[0]
        self.basis = map(lambda poly: str(poly.firstChild.data).strip(),polynomials.getElementsByTagName("poly"))

    def __constructJasObject(self):
        from types import StringType
        from jas import PolyRing, ZZ
        # set up the polynomial ring (Jas syntax)
        if 'hasParameters' in self.__dict__ and self.hasParameters != '':
            #K = 'K.<%s> = PolynomialRing(ZZ)' % self.hasParameters
            #R = K + '; R.<%s> = PolynomialRing(K)' % self.hasVariables
            K = 'PolyRing(ZZ(),"%s")' % self.hasParameters
            R = 'R = PolyRing(%s,"%s")' % (K, self.hasVariables)
            gens = '(one,%s,%s)' % (self.hasParameters, self.hasVariables)
        else:
            #R = 'R.<%s> = PolynomialRing(ZZ)' % (self.hasVariables)
            R = 'R = PolyRing(ZZ(),"%s")' % (self.hasVariables)
            gens = '(one,%s)' % self.hasVariables
        # translate Jas syntax to pure Python and execute
        #exec(preparse(R))
        R = R + "; " + gens + " = R.gens();"
        #print str(R)
        exec(str(R))
        print "R = " + str(R)
        self.jasRing = R;
        # construct polynomials in the constructed ring from
        # the polynomial expressions
        self.jasBasis = []
        for ps in self.basis:
            #print "ps = " + str(ps)
            if type(ps) is StringType and '^' in ps:
               ps = ps.replace('^', '**')
            #exec(preparse("symbdata_ideal = %s" % ps))
            exec("symbdata_poly = %s" % ps)
            self.jasBasis.append(symbdata_poly)
        #print "jasBasis = " + str([ str(p) for p in self.jasBasis])

    # the following functions will all use Jas to
    # calculate metadata
    def jas_hasLengthsList(self):
        """
        This is the implementation of the predicate "sd:hasLengthsList".
        The lengths lists is the sorted list of the number of monomials of
        the generator of the ideal.

        Along with the output, there will also be generated a field
        FROM_JAS_hasLengthsList which can be used to later access the
        data without recalculating. The main reason for this is that the
        SymbolicData properties are converted into field, not getter
        functions. So to have some symmetry, the Jas calculations will
        end up in fields as well.
        """
        try:
            LL = sorted(map(lambda x : len(x.monomials()), self.jasBasis))
            self.FROM_JAS_hasLengthsList = ",".join(map(lambda x: str(x), LL))
        except:
            self.FROM_JAS_hasLengthsList = ''
        return self.FROM_JAS_hasLengthsList

    def jas_hasDegreeList(self):
        """
        This is the implementation of the predicate "sd:hasDegreeList".
        The degree list is the sorted list of the degree of the generator
        of the ideal.

        Along with the output, there will also be generated a field
        FROM_JAS_hasDegreeList which can be used to later access the
        data without recalculating. The main reason for this is that the
        SymbolicData properties are converted into field, not getter
        functions. So to have some symmetry, the Jas calculations will
        end up in fields as well.
        """
        try:
            LL = sorted(map(lambda x : x.degree(), self.jasBasis))
            self.FROM_JAS_hasDegreeList = ",".join(map(lambda x: str(x), LL))
        except:
            self.FROM_JAS_hasDegreeList = ''
        return self.FROM_JAS_hasDegreeList

    def jas_hasVariables(self):
        """
        This is the implementation of the predicate "sd:hasVariables". This
        is actually not needed.
        """
        #K = []
        #DL = map(lambda m : K.extend(map(lambda l : str(l), m.variables())), self.sageBasis)
        K = self.jasRing.ring.vars
        return ",".join(sorted(list(set(K))))

    def jas_homogenize(self, hv):
        """
        Homogenize a basis, which here means actually nothing more than
        homogenizing every element of the basis.
        """
        homBasis = map(lambda x : x.homogenize(hv), self.jasBasis)
        return homBasis

File: examples/sdjas.rb
#
# $Id$
#
# adapted from the Symbolicdata, Sage and JAS Python code
#
# communicate with the SPARQL endpoint (http request)
#import requests
require "net/http" 
# easily handle URLs
#from os.path import split as pathsplit
#from urlparse import urlsplit
require "uri"
# for nice output of dictionaries: json.dumps(dict, indent = 4)
# mostly for debugging reasons (will be removed later)
require "json" 
# parse the sd.ini file
#from ConfigParser import SafeConfigParser
require "configparser"
# parse the xml of the resource files
#from xml.dom.minidom import parse, parseString
require "rexml/document"
# output lists nicely (there might be a better way)
#from textwrap import wrap as textwrap
# not needed
#require "examples/jas"

# Some internal helper functions that are not meant to be
# called by the user

=begin rdoc
    Converts a uri to a name or key by only taking everything
    after the last / or (if present) #.

    Examples:
        http://example.com/test             ->   test
        http://example.com/model#testedBy   ->   testedBy
=end
def _uri_to_name(uri)
    usplit = URI(uri) #urlsplit(uri)
    if usplit.fragment != nil
        return usplit.fragment
    else
        #return pathsplit(usplit.path)[-1]
        return usplit.path.split('/')[-1]
    end
end

=begin rdoc
    Formats a list l to be displayed in a tabular layout. It is
    possible to pass an integer width to the textwrap function.
    The width of the terminal window could be obtained via the
    Python console module. However, since it is not included
    in Jas, we decided not to use it. The default width that
    textwrap uses is set to 70. There might be a better way to
    do this.
=end
def _pprint(l)
    col = max(l.map { |x| x.length } ) + 3 #  [len(x) for x in l]) + 3
    padded = l.map{|x| x.ljust(col) }.join('')
    print textwrap(padded).join('\n')
end

=begin rdoc
    A quick convienience function to retrieve a single value
    of a given triple (object, predicate, ...)

    The parameter sd is a SymbolicData object that contains
    information about the SPARQL endpoint.
=end
def get_value_for_URI(sd, uri, predicate)
    result = nil
    query = "SELECT * WHERE { <#{uri}> <#{predicate}> ?x }"
    begin
        qj = SPARQL.new(sd, query)
        result = qj.json['results']['bindings'][0]['x']['value']
    rescue
        #pass 
    end
    return result
end


# Class definitions start here

=begin rdoc
    Access to the database of ideals as provided by the
    SymbolicData Project (http://symbolicdata.org).
=end
class SymbolicData

      attr_reader :_parser, :url, :sqpath, :sd, :sdhost

=begin rdoc
        The constructor parses the sd.ini file and sets up some variables.
        An optional parameter can be passed to select the SPARQL endpoint
        that should be used. The keywords for different SPARQL endpoints are
        defined in the sd.ini file.

        The default SPARQL endpoint is the one from symbolicdata.org
=end
    def initialize(sparql = 'symbolicdata.org')
        @_sparql = sparql
        @_ideals = nil
        #@_parser = SaveConfigParser.new(['sd.ini','examples/sd.ini'])
        @_parser = ConfigParser.new('examples/sd.ini') if File.file?('examples/sd.ini')
        @_parser = ConfigParser.new('sd.ini') if File.file?('sd.ini')
        @sd = @_parser['symbolicdata']['sd']
        begin
            @url = @_parser['sparql'][@_sparql]
        rescue
            raise ValueError("The SPARQL endpoint referenced by '#{@_sparql}' was not found in the sd.ini file." )
        end
        @sdhost = @_parser['DEFAULT']['sdhost']
        sh, sp, port = @url.partition(':') # hack for subst and GI host error
        @url = @sdhost + sp + port
        @sqpath = @_parser['sparql']['path']
        puts "SymbolicData() initialized"
        puts "url    = " + str(@url)
        #puts "sdhost = " + str(@sdhost)
    end

=begin rdoc
        Returns a Python list of ideals.
=end
    def get_ideals(force_reload = false)
        if @_ideals == nil or force_reload == true
            list_ideals(false, force_reload)
        end
        return @_ideals
    end

=begin rdoc
        Lists all the available ideals.
=end
    def list_ideals(output = true, force_reload = false)
        if @_ideals == nil or force_reload == true
            r = SPARQL.new(self, @_parser.get('queries', 'list_ideals'))
            ids = r.json['results']['bindings']
            @_ideals = ids.map{ |x| _uri_to_name(x['ideal']['value']) }
        end
        if output
            _pputs(@_ideals)
        end
    end

=begin rdoc
        Returns an ideal as a Jas object that is ready to be used by
        Jas.
=end
    def get_ideal(uri)
        return SD_Ideal.new(self, uri).get_ideal();
    end

=begin rdoc
        Returns an internal object that represents the SymbolicData
        database object. (See below for details)
=end
    def get_sd_ideal(uri)
        return SD_Ideal.new(self, uri)
    end

end


=begin rdoc
    This is a 'wrapper' class for SPARQL queries. A class might be
    a slight overkill. It was made with the idea, that one can store
    the query and the result together, to re-evaluate both without
    having to access the server. However, in the end this feature
    was not really needed.
=end
class SPARQL

    attr_reader :json

    def initialize(sd, query, output = 'json')
        @_sd = sd;
        @_query = query;
        @_data = {
            'query' => query,
            'output' => output
        }
        #self.response = requests.get(self._sd.url, params = self._data)
        #puts "url = " + str(self._sd.url)
        uri = URI("http://" + @_sd.url.to_s)
        #puts "uri      = " + uri.to_s
        #puts "uri.host = " + uri.host.to_s
        #puts "uri.port = " + uri.port.to_s
        Net::HTTP.start( uri.host, uri.port ) do |conn|
           #puts "conn = " + str(conn)

           #puts "query = " + str(query)
           #_path = @_sd.sqpath + "?" + URI.encode_www_form(@_data)
           uri.path = @_sd.sqpath
           uri.query = URI.encode_www_form(@_data)
           #puts "uri       = ", uri
           #puts "uri       = " + str(uri)
           #puts "uri.path  = " + uri.path.to_s
           #puts "uri.query = " + uri.query.to_s
           req = Net::HTTP::Get.new(uri.request_uri)
           #puts "req = " + str(req)
           response = conn.request( req );
           if not response.is_a?(Net::HTTPSuccess)
              puts "response = " + response.to_s + "\n"
              raise ValueError, "HTTP GET #{uri} not successful" 
           end

           #head = response.code.to_s + " " + response.msg
           #puts "head = " + str(head) + "\n"
           @text = response.body()
           #puts "body = " + str(@text)
           @json = JSON.load(@text)
           #puts "json = " + str(@json)
        end
    end

end


=begin rdoc
    This class represents a SymbolicData database object. The
    constructor takes a complete URI or a name SUBJ (the latter of which
    will be prefixed with the 'ideal' value from the sd.ini)

    Any triple of the form (SUBJ, PRED, OBJ) will yield a field PRED*
    for the SD_Ideal object with the value OBJ, where PRED* is the
    ending piece of PRED URI as defined by the function _uri_to_name()

    A SPARQL endpoint is needed. As a future improvement, it could be
    nice to directly parse an RDF in a convienient serialization.
=end
class SD_Ideal

    attr_reader :basis

=begin rdoc
        sd is a SymbolicData object, the name can be a complete URI or shortened
        name as defined by _uri_to_name(). The latter will be prefixed with the
        'ideal' value from the sd.ini. Namespaces like "sd:Wu-90" are not
        (yet) supported.

        Appart from retrieving the information from the SPARQL endpoint, the
        resource data (XML files) is needed as well. While the SPARQL endpoint
        can be substituted by another SPARQL endpoint, the links to the resource
        files are 'hard-coded' into the RDF data. The possibility to use a
        (possibly 'hand-filled') cache will be included in the next update.
=end
    def initialize(sd, name)
        @dict = {} # mimic python
        @_sd = sd
        # quick test, if the given name already is an uri
        if name[0,7] == 'http://'
            @uri = name
        else
            @uri = @_sd._parser["symbolicdata"]["ideal"] + name
        end

        @dict["hasXMLResource"] = false
        @dict["hasLengthsList"] = ''
        @dict["hasDegreeList"] = ''
        @dict["hasParameters"] = ''

        @basis = []

        # we set up the query to get all predicate values
        # of the URI/polynomial system/ideal
        query = "
            PREFIX sd: <#{@_sd.sd}>
            SELECT ?p ?o WHERE {
                <#{@uri}> ?p ?o
            }"
        #puts "query = " + query.to_s + "\n" 
        @_request = SPARQL.new(@_sd, query)

        if @_request.json['results']['bindings'].size == 0
            raise ValueError("No data found for <#{@uri}>.\nMaybe the name was misspelled or the SPARQL endpoint is unavailable.")
        end

        # append the keys to the @dict.
        for t in @_request.json['results']['bindings']
            uri = t['p']['value']
            obj = t['o']['value']
            @dict[_uri_to_name(uri)] = obj
        end

        # Next we need a resource file with the actual expressions that are
        # used to generate the ideal.
        #
        # There are four cases that need to be dealt with
        #     (1) the ideal is constructed direclty
        #         from an IntPS with related XML resource
        #     (2) the ideal is a flat variant of another
        #         ideal
        #     (3) the ideal is obtained by homogenizing
        #         another ideal
        #     (4) the ideal is obtained by parameterizing another
        #         ideal
        # Please note: While it might seem that only one of (2) and (4)
        # should be included, both are needed to map the actual history
        # of how these ideals were obtained.

        # case 1
        if @dict.include?( 'relatedPolynomialSystem' )
            link = get_value_for_URI(@_sd, @dict["relatedPolynomialSystem"], @_sd.sd+'relatedXMLResource')
            __addXMLResource(link)
            @dict["hasXMLResource"] = true
            #puts "relatedPolynomialSystem " + str(name)
        end

        # case 2
        if @dict.include?( 'flatten' )
            parent_name = @dict["flatten"]
            parent = SD_Ideal.new(@_sd, parent_name)
            @variablesCSV = @dict["hasVariables"]
            @variables = @variablesCSV.split(",").map{ |x| x.to_s.strip() }
            @basis = parent.basis
            #puts "flatten " + str(parent_name) + ", name = " + str(name)
        end

        # case 3
        if @dict.include?( 'homogenize' )
            parent_name = @dict["homogenize"]
            if @dict.include?( 'homogenizedWith' )
                hv = @dict["homogenizedWith"]
                parent = SD_Ideal.new(@_sd, parent_name)
                @variablesCSV = parent.variablesCSV + "," + hv
                @variables = parent.variables
                @variables.append(hv)
                @basis = parent.jas_homogenize(hv)
            #puts "homogenize " + str(parent_name) + ", name = " + str(name)
            end
        end

        # case 4
        if @dict.include?( 'parameterize' )
            parent_name = @dict["parameterize"]
            parent = SD_Ideal.new(@_sd, parent_name)
            @variablesCSV = @dict["hasVariables"]
            @variables = @variablesCSV.split(",").map{ |x| x.to_s.strip() }
            @basis = parent.basis
            #puts "parameterize " + str(parent_name) + ", name = " + str(name)
        end

        # now we got the variables, the parameters and
        # the strings/expressions for the polynomials
        __constructJasObject()
    end


=begin rdoc
        Return the ideal as a Jas objects.
=end
    def get_ideal()
        #return ideal(@sageBasis)
        return @jasRing.ideal("",@jasBasis)
    end

=begin rdoc
        Fill internal objects.
=end
    def __addXMLResource(link)
        #xml = requests.get(link).text
        #puts "link_xml = " + str(link)
        #url = link[0:23]
        path = link[23,link.length-23] # hack for lost domain
        #puts "url = " + str(url)
        #url = @_sd.url[:-5]
        url = URI("http://" + @_sd.sdhost.to_s)
        #puts "url = " + str(url)

        xml = nil
        Net::HTTP.start( url.host, url.port ) do |conn|
           #conn = httplib.HTTPConnection(url)
           #puts "conn = " + str(conn)
           #puts "path = " + str(path)
           url.path = path
           #puts "path = " + str(url.request_uri)
           req = Net::HTTP::Get.new(url.request_uri)
                 #conn.request("GET", path );
           response = conn.request( req );
           if not response.is_a?(Net::HTTPSuccess)
              puts "response = " + response.to_s + "\n"
              raise ValueError, "HTTP GET #{url} not successful" 
           end
           #puts "head = " + response.code.to_s + " " + response.msg + "\n"
           xml = response.body();
        end
        puts _uri_to_name(link).to_s + " = " + xml.to_s

        #xmlTree = parseString(xml)
        xmlTree = REXML::Document.new(xml)

        # Code snipped borrowed from Albert Heinle
        if xmlTree.elements.to_a("*/vars").empty? # Check, if vars are there
            raise ValueError("The given XMLString does not contain variables for the IntPS System")
        end
        if xmlTree.elements.to_a("*/basis").empty? # Check, if we have a basis
            raise ValueError("The given XMLString does not contain a basis for the IntPS System")
        end
        # -------------------- Input Check finished --------------------
        # From here, we can assume that the input is given correct
        @variablesCSV = xmlTree.elements.to_a("*/vars")[0].text
        #puts "@variablesCSV = " + @variablesCSV.to_s
        @variables = @variablesCSV.split(",").map{|x| x.to_s.strip() }
        #polynomials = xmlTree.elements.to_a("*/basis")[0]
        @basis = xmlTree.elements.to_a("*/basis/poly").map{ |poly| poly.text.to_s.strip() }
        #puts "@basis = " + @basis.to_s
    end

=begin rdoc
=end
    def __constructJasObject()
        #from jas import PolyRing, ZZ
        #require "jas"
        # set up the polynomial ring (Jas syntax)
        if @dict.include?('hasParameters') and @dict['hasParameters'] != ''
            #K = 'K.<%s> = PolynomialRing(ZZ)' % @hasParameters
            #R = K + '; R.<%s> = PolynomialRing(K)' % @hasVariables
            kk = 'PolyRing.new(ZZ(),"%s")' % @dict['hasParameters']
            rr = 'rr = PolyRing.new(%s,"%s")' % [ kk, @dict['hasVariables']]
            gens = 'one,%s,%s' % [@dict['hasParameters'], @dict['hasVariables']]
        else
            #R = 'R.<%s> = PolynomialRing(ZZ)' % (@hasVariables)
            rr = 'rr = PolyRing.new(ZZ(),"%s")' % @dict['hasVariables']
            gens = 'one,%s' % @dict['hasVariables']
        end
        # translate JAS syntax to pure Python and execute
        #exec(preparse(R))
        rr = rr + "; " + gens + " = rr.gens();"
        #puts "rr = " + str(rr)
        myb = binding
        eval(rr.to_s, myb)
        @jasRing = rr;
        #eval(gens + " = rr.gens();", myb)
        # construct polynomials in the constructed ring from
        # the polynomial expressions
        @jasBasis = []
        for ps in @basis
            #puts "ps = " + str(ps)
            if ps.is_a? String and ps.include?('^')
               ps = ps.sub('^', '**')
            end
            #exec(preparse("symbdata_ideal = %s" % ps))
            eval("symbdata_poly = %s" % ps, myb)
            @jasBasis.push(myb.eval("symbdata_poly"))
        end
        #puts "jasBasis = " + str(@jasBasis)
    end


    # the following functions will all use Jas to
    # calculate metadata

=begin rdoc
        This is the implementation of the predicate "sd:hasLengthsList".
        The lengths lists is the sorted list of the number of monomials of
        the generator of the ideal.

        Along with the output, there will also be generated a field
        FROM_JAS_hasLengthsList which can be used to later access the
        data without recalculating. The main reason for this is that the
        SymbolicData properties are converted into field, not getter
        functions. So to have some symmetry, the Jas calculations will
        end up in fields as well.
=end
    def jas_hasLengthsList()
        begin
            ll = @jasBasis.map{|x| x.size() }.sort()
            @FROM_JAS_hasLengthsList = ll.map{|x| x.to_s}.join(",") 
        rescue
            @FROM_JAS_hasLengthsList = ''
        end
        return @FROM_JAS_hasLengthsList
    end

=begin rdoc
        This is the implementation of the predicate "sd:hasDegreeList".
        The degree list is the sorted list of the degree of the generator
        of the ideal.

        Along with the output, there will also be generated a field
        FROM_JAS_hasDegreeList which can be used to later access the
        data without recalculating. The main reason for this is that the
        SymbolicData properties are converted into field, not getter
        functions. So to have some symmetry, the Jas calculations will
        end up in fields as well.
=end
    def jas_hasDegreeList()
        begin
            ll = @jasBasis.map{|x| x.degree() }.sort
            @FROM_JAS_hasDegreeList = ll.map{|x| x.to_s}.join(",")
        rescue
            @FROM_JAS_hasDegreeList = ''
        end
        return @FROM_JAS_hasDegreeList
    end

=begin rdoc
        This is the implementation of the predicate "sd:hasVariables". This
        is actually not needed.
=end
    def jas_hasVariables()
        #K = []
        #DL = map(lambda m : K.extend(map(lambda l : str(l), m.variables())), @sageBasis)
        kk = @jasRing.ring.vars
        return sorted(list(set(kk))).join(",")
    end

=begin rdoc
        Homogenize a basis, which here means actually nothing more than
        homogenizing every element of the basis.
=end
    def jas_homogenize(hv)
        homBasis = @jasBasis.map{ |x| x.homogenize(hv) }
        return homBasis
    end

end



to_skip = <<TOSKIP

TOSKIP
